Sources/IfTag.swift

```diff
@@ -100,21 +107,74 @@ final class IfExpressionParser {
   let tokens: [IfToken]
   var position: Int = 0
 
-  init(components: [String], tokenParser: TokenParser) throws {
-    self.tokens = try components.map { component in
-      if let op = findOperator(name: component) {
-        switch op {
-        case .infix(let name, let bindingPower, let cls):
-          return .infix(name: name, bindingPower: bindingPower, op: cls)
-        case .prefix(let name, let bindingPower, let cls):
-          return .prefix(name: name, bindingPower: bindingPower, op: cls)
+  private init(tokens: [IfToken]) {
+    self.tokens = tokens
+  }
+  
+  static func parser(components: [String], tokenParser: TokenParser) throws -> IfExpressionParser {
+    try IfExpressionParser.validateBracketsBalance(components: components)
+    return try IfExpressionParser(components: components, tokenParser: tokenParser)
+  }
+
+  static func validateBracketsBalance(components: [String]) throws {
+    guard try components.reduce(0, {
+      var bracketsBalance = $0
+      if $1 == "(" { bracketsBalance += 1 }
+      else if $1 == ")" { bracketsBalance -= 1 }
+      if bracketsBalance < 0 { throw TemplateSyntaxError("unbalanced brackets") }
+      return bracketsBalance
+    }) == 0 else {
+      throw TemplateSyntaxError("unbalanced brackets")
+    }
+  }
+  
+  private init(components: [String], tokenParser: TokenParser) throws {
+    var parsedComponents = [String]()
+    self.tokens = try components.enumerated().flatMap { index, component in
+      if component == "(" {
+        let (expression, parsed) = try IfExpressionParser.subExpression(
+          from: index + 1,
+          components: components,
+          tokenParser: tokenParser
+        )
+        parsedComponents.append(contentsOf: parsed)
+        return .subExpression(expression)
+      }
+      else if component == ")" {
+        parsedComponents.append(component)
+        return nil
+      } else if index >= parsedComponents.count {
+        parsedComponents.append(component)
+        if let op = findOperator(name: component) {
+          switch op {
+          case .infix(let name, let bindingPower, let cls):
+            return .infix(name: name, bindingPower: bindingPower, op: cls)
+          case .prefix(let name, let bindingPower, let cls):
+            return .prefix(name: name, bindingPower: bindingPower, op: cls)
+          }
         }
+        return .variable(try tokenParser.compileFilter(component))
+      } else {
+        return nil
       }
-
-      return .variable(try tokenParser.compileFilter(component))
     }
   }
 
+  static func subExpression(from index: Int, components: [String], tokenParser: TokenParser) throws -> (Expression, [String]) {
```

Comment:
Wouldn't it be better performance and memory-wise to use an ArraySlice here for components, to prevent rebuilding a new sub-array from the full array on every recursion level? After all, that's exactly what ArraySlices are designed for 😉

That would allow you to get rid of 'Array(…)` wrapping 2 lines below too.
Of course you'd need to replace your `components.enumerated()` on line 133 by `zip(components, components.indices)` instead, which is generally better anyway (see Soroush's article about enumerated) because not all startIndices are 0.

Reply:
Can't say anything about performance, but the only thing it will really change is move `suffix` to caller site, because `IfExpressionParser` accepts `Array`. But it allows to remove one function parameter, so I'll make the change still.

Speaking of enumeration I don't see what `zip` will improve in this particular case. I find `enumerated` use to be more clear here.

Reply:
It's not about readability but about correctness (see Soroush's article), especially if we move to use slices (because they're the type to use when you recurse on subexpression, as they avoid a ton of memory reallocation, and that's exactly what we're doing here so that's a textbook use case) and enumerated will lead to wrong indexes with slices unlike zip (as explained by Soroush)

Reply:
Ok, if this type only accepts `ArraySlice` in constructor it makes sense 👍